processing/src/main/java/org/apache/druid/query/filter/SelectorDimFilter.java

```diff
@@ -71,7 +73,9 @@ public SelectorDimFilter(
   )
   {
     Preconditions.checkArgument(dimension != null, "dimension must not be null");
-
+    this.longPredicate = makeLongPredicateSupplier().get();
```

Comment:
The semantic is changed: now predicates are not lazy. Here, you cannot use memoize() as in other classes - you should more scrupulously follow the "golden" examples forged in #6662.

Reply:
Hi Leventov, I am not sure if I understand this case correctly : basically the original code is saying I want 3 private predicates  only be initialized once when  SelectorDimFilter::toFilter() method has been called  +  DruidPredicateFactory::makeDoublePredicate() has been called, is it correct? My question is since DruidPredicateFactory is final(already thread safe), the whole purpose of the old logic is just not init the predicate twice ,all I need to do is change them to final...
   ` @Override
  public Filter toFilter()
  {
    if (extractionFn == null) {
      return new SelectorFilter(dimension, value, filterTuning);
    } else {

      final DruidPredicateFactory predicateFactory = new DruidPredicateFactory()
      {
        @Override
        public Predicate<String> makeStringPredicate()
        {
          return Predicates.equalTo(value);
        }

        @Override
        public DruidLongPredicate makeLongPredicate()
        {
          initLongPredicate();
          return longPredicate;
        }

        @Override
        public DruidFloatPredicate makeFloatPredicate()
        {
          initFloatPredicate();
          return floatPredicate;
        }

        @Override
        public DruidDoublePredicate makeDoublePredicate()
        {
          initDoublePredicate();
          return druidDoublePredicate;
        }
      };
      return new DimensionPredicateFilter(dimension, predicateFactory, extractionFn, filterTuning);
    }
  }
`

Reply:
> DruidPredicateFactory is final(already thread safe)

DruidPredicateFactory is not thread safe, as far as I can tell.

You cannot initialize predicates eagerly since it may not be possible: e. g. if the `value` is `"1.0"`, it will not be parseable as long in `createLongPredicate()`. The method will not throw because it uses `tryParseLong()`, but nevertheless this is fragile.

The problem with the current code is that it allows spurious null returns of predicates due to double-read of a non-volatile field, as described [here](https://github.com/code-review-checklists/java-concurrency#safe-local-dcl), see [here](https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#wishful-hb-actual) for more theoretical foundations.

You can fix this by either applying proper double-checked locking, or using [*racy single check initialization*](https://github.com/code-review-checklists/java-concurrency#lazy-init-benign-race).

BTW, also makes sense to cache `makeStringPredicate()`.